### PR TITLE
Add Krea 2 LoRA training (mflux-train)

### PR DESCRIPTION
**Pull request by fxd0h**
## Krea 2 LoRA training

This adds LoRA fine-tuning for [Krea 2](https://github.com/krea-ai/krea-2) via `mflux-train`, on top of the Krea 2 inference port in #456.

It implements the existing `TrainingAdapter` protocol (same shape as the `z_image` / `flux2` / `ernie_image` adapters), so it plugs into the current training runner, trainer, dataset cache, and LoRA injection without touching the shared training loop.

```jsonc
// train.json
{
  "model": "krea-2-raw",
  "data": "images/",
  "quantize": 8,
  "max_resolution": 512,
  "training_loop": { "num_epochs": 16, "batch_size": 1, "timestep_low": 3, "timestep_high": 18 },
  "optimizer": { "name": "AdamW", "learning_rate": 2e-4 },
  "lora_layers": { "targets": [
    { "module_path": "transformer_blocks.{block}.attn.to_q", "blocks": { "start": 0, "end": 28 }, "rank": 16 }
    // ... to_k, to_v, to_out.0, ff.gate, ff.up, ff.down
  ]}
}
```
```bash
mflux-train --config train.json
```

### What is included
- `Krea2TrainingAdapter` implementing `TrainingAdapter`:
  - `encode_data`: VAE-encode + Krea 2 latent packing, Qwen3-VL layer-tap text encode (12 layers) + mask, 3-axis RoPE position ids. Captions may be plain text or the Ideogram JSON schema (the natural-language description is extracted).
  - `predict_noise`: a single transformer forward, no CFG. Flow-matching velocity: the trainer noises with `z = (1 - sigma) * clean + sigma * noise`, so the target velocity `noise - clean` is the model's native prediction at `timestep = sigma`; the adapter feeds `t = sigma` and returns the velocity unchanged.
  - QLoRA over a q8 base, saving / loading the LoRA under the `transformer.*` prefix (the same namespace the inference LoRA loader matches).
- Optional gradient checkpointing in `Krea2Transformer`'s block loop (off by default, inference unaffected; the adapter enables it so the ~12B activation graph fits in memory).
- Krea 2 selection wired into the training runner.

### Validation
Trained a small style LoRA on Krea 2 Raw (q8) end to end on an M5 Max: 6 images, 96 steps, trigger token `bluxyz`. The resulting LoRA loads through the existing Krea 2 LoRA path (`All LoRA weights applied successfully`) and visibly applies the learned style at inference. Same prompt and seed, base vs LoRA:

Base vs LoRA, same prompt and seed

Depends on #456 (Krea 2 inference): the training adapter imports the Krea 2 model, so this branch is stacked on it and its diff currently shows both commits. Once #456 lands, this reduces to just the training change (the `Add Krea 2 LoRA training adapter` commit). Reviewing #456 first keeps each change to one logical unit.

### CODE REVIEW

**Comment by ianscrivener** [comment]
@fxd0h can you please can push the Krea2 MFlux model to Huggingface - Q6 or Q8. I'm keen to test your PR & Krea2.

**Comment by plz12345** [comment]
> @fxd0h can you please can push the Krea2 MFlux model to Huggingface - Q6 or Q8. I'm keen to test your PR & Krea2.

https://huggingface.co/MLXBits/krea-2-mlx-q8/

**Comment by ianscrivener** [comment]
No luck weith https://huggingface.co/MLXBits/krea-2-mlx-q8/ 
image

**Comment by plz12345** [comment]
> No luck weith https://huggingface.co/MLXBits/krea-2-mlx-q8/ image

Try it with my implementation in #453

**Comment by ianscrivener** [comment]
thanks @plz12345 - working now
image

**Comment by fxd0h** [comment]
superseded by #462, which adds Krea 2 Raw + the training on top of the merged #453 (this branch was built on the earlier inference). closing.
